src/language-js/printer-estree.js

```diff
@@ -3094,12 +3094,29 @@ function printPathNoParens(path, options, print, args) {
       return "unknown";
     case "TSVoidKeyword":
       return "void";
-    case "TSAsExpression":
-      return concat([
-        path.call(print, "expression"),
-        " as ",
-        path.call(print, "typeAnnotation")
+    case "TSAsExpression": {
+      const typeText = options.originalText.slice(
+        options.locStart(n.typeAnnotation),
+        options.locEnd(n.typeAnnotation)
+      );
+      // Defer to expansion logic for objects, arrays, and annotations wrapped in parentheses.
+      const mightBeMultiline = /^[({[]/.test(typeText.trim());
```

Comment:
Thoughts on a better way to check for this? If we always insert a newline here, the formatting isn't as nice. i.e.

```ts
const value2 = thisIsAReallyLongIdentifier as (SomeInterface | SomeOtherInterface);
```
becomes
```ts
const value2 = thisIsAReallyLongIdentifier as 
  (SomeInterface | SomeOtherInterface);
```
rather than
```ts
const value2 = thisIsAReallyLongIdentifier as (
  | SomeInterface
  | SomeOtherInterface
);
```

Reply:
```
const value2 = thisIsAReallyLongIdentifier as 
  (SomeInterface | SomeOtherInterface);
```
looks better to me

Reply:
And especially this
**Prettier pr-6450**
[Playground link](https://deploy-preview-6450--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuc0DOMAEA3AhgGwFc4AmTAXkxgAsBLNASTQEEAlOA-ATwBloBzBgBMEMWgDNacAE6ZcaTAGUIAWzgNYM8bjBwAPO068BAFS4AHOAAVc03GvjSAfAG4QAGhARzY9MlC20hAA7jbSCGjIILjYELRCHiAARnZgANZwMIrmOrRQ-Mgw0sSe1DAq+ADqdPBoObqKEbRi2M1cUWBokZ55aDIwVnb8KrjI2vh9ngBWaAAeAEKpGVn2cDx5cGMEkyAzs4p5-PhwAIqEEPBbE3CeOdJ90lEwFnBoYNK0PonmH7CV8TRkAAOAAMtyCfUqdnMUR+rxk2E2ngAjud4INvJEUPIALRQOBwEQJTzhVG0cKDXDDUZIcY7PoqWiFYo3EBoQ7HM4XTa07asmC4JL-ISApAkTxFXC0fCHADCqhGUSg0CRIEIfRMgqxdNZ2GIGhEsEU70+MGYUCEimexyufQAvnagA)
```sh
--parser typescript
```

**Input:**
```tsx
const value2 = thisIsAReallyLongIdentifier as SomeInterface<ReallyLongTypeParameter>;
```

**Output:**
```tsx
const value2 = thisIsAReallyLongIdentifier as
  SomeInterface<ReallyLongTypeParameter>;

```

is way better than this

**Prettier 1.18.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEA3AhgGwFc4AmTAXkxgAsBLNASTQEEAlOA-ATwBloBzBgBMEMWgDNacAE6ZcaTAGUIAWzgNYM8bjBwAPMrUB5GjL5RBm6dt0A+ANwgANCAgAHMemShc06RAB3AAVfBDRkEFxsCFohZxAAI2kdAGs4GEU3HVoLZBhpYhdqGBV8AHU6eDQs3UUw2jFsBq4IsDRwlxy0GRgg5P4VXGRtfG6XACs0AA8AIWSwNIzcNR4cuGGCMZBJqcUc-nw4AEVCCHgN0bgXLOlu6QiYLjc4NDBpWg94t3fYMtiaZAADgADNd-N0ysk3BFvi8ZNh1i4AI6neB9dzhFDyAC0UDgcBEcRc0jgKNoJL6uAGQyQIy23RUtDyBSuIDQ+0OJzO61pm1ZMFwCT+QgBSBILnyuFo+H2AGFVIMIlBoIiQIRugAVQWYulwAC+eqAA)
```sh
--parser typescript
```

**Input:**
```tsx
const value2 = thisIsAReallyLongIdentifier as SomeInterface<SomeOtherLongInterface>;
```

**Output:**
```tsx
const value2 = thisIsAReallyLongIdentifier as SomeInterface<
  SomeOtherLongInterface
>;

```

Reply:
Thanks for the input!

Re the parentheses, I think Prettier should still format it correctly, whether they're required or not.

Re the second comment, I think formatting generics should be treated separately from this PR (and none of the changes here affect that, since it only checks the first token after the `as` operator), to make sure it's as small a patch as possible.

Reply:
> I think formatting generics should be treated separately from this PR

It's non-wrapping of `as` that causes #5554. So this PR is going to fix it.

Reply:
As for the parentheses, I'm just saying it's probably a bug that Prettier doesn't remove them. But it's out of scope for this PR, sure.